torch/distributed/distributed_c10d.py

```diff
@@ -361,7 +362,8 @@ def init_process_group(backend,
                        world_size=-1,
                        rank=-1,
                        store=None,
-                       group_name=''):
+                       group_name='',
+                       deviceId=None):
```

Comment:
I would like to pull in more discussion on this.  cc @pritamdamania87 @zhaojuanmao @rohan-varma @osalpekar 

Question, what if the user would like to create a process to work on two different GPUs? E.g., rank0 on GPU0 and GPU1 and rank1 on GPU2 and GPU3? And what if users created process group for CPU only applications?

Reply:
I am trying to understand more about this change. If we can replace the following `barrier()` by Store-based synchronization, will it be possible to drop this argument?

https://github.com/pytorch/pytorch/blob/220b91660f5b1a098b7238d8f6f5ae95048685cd/torch/distributed/distributed_c10d.py#L477

Reply:
> I am trying to understand more about this change. If we can replace the following barrier() by Store-based synchronization, will it be possible to drop this argument?

I don't think this will resolve the problem entirely since the user can still call `barrier()` manually after `init_process_group`.

Reply:
> Question, what if the user would like to create a process to work on two different GPUs? E.g., rank0 on GPU0 and GPU1 and rank1 on GPU2 and GPU3? And what if users created process group for CPU only applications?

I think if we do want to support this option it probably needs to support a device_list here. The argument type should probably be `List[torch.device]` so it can cover CPU as well.

Reply:
> > I am trying to understand more about this change. If we can replace the following barrier() by Store-based synchronization, will it be possible to drop this argument?
> 
> I don't think this will resolve the problem entirely since the user can still call `barrier()` manually after `init_process_group`.

Yes, the main issue is calling barrier inside init_process_group. By adding device or tensor parameter, the user is forced to use it, though its not user intention.
Better option is not to call "all-reduce" barrier during init_process_group. I did not observe any NCCL call for the init_process group. In that case, we can replace GPU based barrier to Python based or a regular barrier. @mrshenli mentioned about Store-based synchronization as replacement.

Reply:
> > Question, what if the user would like to create a process to work on two different GPUs? E.g., rank0 on GPU0 and GPU1 and rank1 on GPU2 and GPU3? And what if users created process group for CPU only applications?
> 
> I think if we do want to support this option it probably needs to support a device_list here. The argument type should probably be `List[torch.device]` so it can cover CPU as well.

Again, the issue is specific to NCCL backend. Even if CPU devices are added, it will be redundant.
As I mentioned above, the bug is due to barrier call inside init_process_group. If we can replace the barrier inside init_process_group with similar function then probably we do not need the PR changes.

Reply:
Could we go  with this approach for a general solution to the issue in `barrier()` (probably change it to list as @pritamdamania mentioned), and then instead of providing an additional flag in `init_process_group`, always just default to store-based synchronization instead of a barrier in init_process_group?

Reply:
@rohan-varma @mrshenli @jaglinux I have a PR to replace the barrier with a store based one for init_process_group: https://github.com/pytorch/pytorch/pull/49419. Maybe we can update this PR to include the device_list for barrier?

Reply:
Agreed, the PR #49419 fixes the original issue. I will modify this PR to add device list and remove the init_process_group changes. Do we need to cover CPU cases ? since barrier allreduce is specific to NCCL and GPU.
By the way, are the test cases activated to test multi GPU's per process ?